### ISSUE TITLE
Custom DL path SAF issue

### ISSUE THREAD

**Issue by boognish-rising** [recloudstream/cloudstream#480]
### Steps to reproduce

100% reproducible with my hardware and I'm reasonably sure the same will be true of any Fire device that connects to a TV and/or is the TV's OS (Amazon FireTV or any other brand with FireOS at the helm). Confirmed using Hisense U6HF 

1. Settings→General→Download Path→Custom+Apply
2. This will bring up Recent folder in System Access Framework (cough🖕SAF cough). If you do anything other than immediately clicking enter, you're SOL and as far as I can tell, you're only option is to click back and try again. If you do click enter immediately, it opens the "hamburger" menu (i.e., ≠3 dots in the upper right but the 3 horizontal lines in the upper left. I feel like it's been many Android versions since a menu was referred to as a hamburger menu) where you can _see_ DLs, Internal Storage, Storage Devices, etc. but again, as far as I can tell, there is no way to select anything other than "Recent" but even then, you're really just dismissing the menu options. 

That's the end of the line unless you've got a mouse or other input device connected via OTG, in which case:

3. This requires much patience, as a cursor is not visible on-screen, but eventually one can select/open any option listed in the menu above. (lower DPI=better) 
4. Trouble is, selecting/clicking "Select" to confirm the custom DL path, as far as I can tell, is **impossible** 
(It would appear that, at least with FireOS and probably other Fire devices but possibly also AndroidTVs and perhaps even GoogleTVs, that _visible_ mouse cursors are not supported on system screens, despite support being present for the input device itself, however, on one such system screen where all the same elements were present (invisible but present cursor) I was able to select "OK/Select/Agree" that was located in exactly the same place as "Select" is in SAF as mentioned above on the first attempt (not a skill I asked for or want but I guess I've become proficient at navigating blindly via mouse)

### Expected behavior

Ability to select a custom DL path without breaking things and ultimately failing anyway

### Actual behavior

Inability to select any DL path other than the 2 paths provided above "Custom" (/storage/emulated/0/Download & /storage/emulated/0/Cloudstream)

### Cloudstream version and commit hash

4.0.1 ded96dd9 & 0d431fd

### Android version

Android 9 (FireOS 7.6.3.3)

### Logcat

```java
[logcat.txt](https://github.com/recloudstream/cloudstream/files/11648533/logcat.txt)
```

### Other details

Sorry for my long-winded nature. I try to provide any/all relevant details when reporting issues that I'm unable to work out myself. Also, while there were some slightly related issues, I looked through all open issues and didn't see any that even mention the crux of mine (however I didn't look through closed issues, assuming this wouldn't be among them since it is still very much present/persistent). Lastly, as reflected by app version(s), I have both beta and stable installed and confirmed the issue is present with both but in the interest of not muddying the waters (not to mention privacy and unnecessarily involving/mentioning extensions/repos/providers and all their related data), I've only included logs from stable, as it is still mostly unconfigured and without any added repos, as their presence seems to have absolutely no bearing on reproducibility. 

Edit - as alluded to, I don't actually think this is a Cloudstream bug, which is why I wrestled with whether to label it as a bug or an enhancement but settled on bug and despite it not being CS3's bug per se, decided it would be prudent to submit anyway, as even a workaround would be better than no ability to select a custom DL path at all (e.g., if possible, providing additional preconfigured paths to select from, thereby avoiding SAF and it's BS entirely). However, if anyone has found another workaround that can already be used/requires no dev involvement, feel free to close the issue but do please make specific mention of said workaround so others can benefit and more importantly, never have to invoke SAF, preferably ever again... Bc 🖕SAF

### Acknowledgements

- [X] I have searched the existing issues and this is a new ticket, **NOT** a duplicate or related to another open issue.
- [X] I have written a short but informative title.
- [X] I have updated the app to pre-release version **[Latest](https://github.com/recloudstream/cloudstream/releases)**.
- [ ] If related to a provider, I have checked the site and it works, but not the app.
- [X] I will fill out all of the requested information in this form.

**Comment by boognish-rising** [comment]
Uhhh... I didn't mention any provider, Sorastream or otherwise and specifically mentioned that they have no bearing. I even specifically went out of my way to not mention any, even by way of logs, so thanks for your input, bot, but go home... You're drunk

**Comment by Blatzar** [comment]
Not possible to have more pre-configured paths, even having the downloads folder available is not easy with scoped storage.
Download AnExplorer to get a decent file picker, which can select a custom path properly.

**Comment by boognish-rising** [comment]
Thanks for the response, @Blatzar. I'm not sure if I'm reading you wrong but are you saying that there's a way to launch AnExplorer (or presumably any 3rd party file manager) instead of SAF in the scenario I described? Because that would be **ideal**. I had tried changing "Open Supported Links" to "Open in this app" but it made no difference. SAF was/is still launched when I attempt to select a custom DL path. This was using my go-to FM, X-plore (I think I tried Ghost Commander, too) but if for reasons unbeknownst to me, AnExplorer _does_ work when X-plore _doesn't_, I'd install it immediately. What's one more FM? 😂 I'm surprised I'm unfamiliar with it, as I'm a bit of an app whore... thought that I'd used every FM worth a damn but AnExplorer is not ringing any bells. 


•Possibly relevant, X-plore's supported ~~links~~ link is: * 

∘Possibly irrelevant but I noticed an hour or so ago that there's suddenly a 3rd "pre-configured" path option above "Custom" (/storage/emulated/0/Android/data/com.lagradost.cloudstream3.prelease/files). Any idea where that might've come from or if whatever means brought it about can be replicated with the path I'm trying to set(USB storage device)? The only thing I can think of is that somehow capturing the logs triggered it but I doubt it, as they were saved elsewhere.

•Last question and it might be a dumb one but are there _any_ other ways to set the DL path? I don't know the first thing about the programming side of your app/what's under the hood but I'd imagine that it would need to be outside of the GUI, i.e., via terminal, etc.

∘None of this would be an issue if Smart TVs (dongles are no better) provided more internal storage. I feel like my Hisense is on the high end of the spectrum at 12GB provided but that makes zero diff when a _single_ 4K feature length with DOLBY, DTS, or whatever sound profile and codecs and whatnot that you're into weighs in @ 50GB+. K done griping. 

Thanks for your help, I think I'll close this unless you think it should remain open. Though, we _are_ able to continue commenting after it's closed if necessary, correct? For example, if you replied to the questions I posed above in such a way that I may have follow up stuff, etc. Either way, your call. 

PS - ran into almost identical scenario with another app, same result/observed behavior etc

**Comment by Blatzar** [comment]
On any competent android system it should make AnExplorer available to use, either by a prompt when you click apply in cs3, or within the system file manager. I cannot help you beyond that regarding on how to make it open the correct app. Something is funky with your TV.

> •Last question and it might be a dumb one but are there any other ways to set the DL path? I don't know the first thing about the programming side of your app/what's under the hood but I'd imagine that it would need to be outside of the GUI, i.e., via terminal, etc.

No*
Cs3 basically asks the OS to select a folder and it returns a content:// path. This also authenticates the app to use that path, which makes it impossible to skip this step. It does not simply ask for a path like _/sdcard/MyCustomFolder_ it also makes it possible for the app to use that folder. Hardcoding _/sdcard/MyCustomFolder_ would therefore not work. Therefore you cannot restore a cs3 backup with your custom folder selected to 'trick' cs3 into downloading where you want, you need to pick a path.

*The only real way to skip this app wise would be to do the same as a file manager app and ask for permissions for everything, not something we will do to fix a niche problem with a niche feature. 

Basically an OS problem, not app problem, and certainly nothing we can do anything about.
If you want to tinker with it yourself, look here: https://github.com/recloudstream/cloudstream/blob/master/app/src/main/java/com/lagradost/cloudstream3/ui/settings/SettingsGeneral.kt#L126-L152

> ∘Possibly irrelevant but I noticed an hour or so ago that there's suddenly a 3rd "pre-configured" path option above "Custom" (/storage/emulated/0/Android/data/com.lagradost.cloudstream3.prelease/files). Any idea where that might've come from or if whatever means brought it about can be replicated with the path I'm trying to set(USB storage device)? The only thing I can think of is that somehow capturing the logs triggered it but I doubt it, as they were saved elsewhere.

If you can move cs3 itself to your external storage the cs3 app's 'private' storage will be on your external storage too, making it possible to select as a pre-configured path afaik. Worth checking out.

**Comment by boognish-rising** [comment]
Thanks again. I took some of your recommendations and incorporated some of my own that our back and forth inspired and ultimately I ended up with a good compromise (your suggestion to move CS3 to external storage so that it & internal/private storage  would be one in the same was looking very promising and is a great suggestion actually, however, for some reason that I have yet to figure out, my storage device likes to periodically unmount and remount itself and apps like CS3 & any other that I might use for streaming / viewing content don't particularly care for these periodic brownouts and regardless of their preference, it's not conducive for content consumption. Honestly, it's not conducive to anything, but it's especially inconvenient for this purpose. So until I figure that out, this solution will have to sit on a back burner). I discovered that while it is extremely tedious and frustrating, the lack of a visual indicator/representation of the mouse cursor (physical or emulated mouse, doesn't matter), it is there and can in fact reach the "Select" button, in contrast to what I had written/thought previously. It takes a tremendous amount of patience but armed with this new insight, I downloaded [CIFS Documents Provider](https://github.com/wa2c/cifs-documents-provider), which somehow tricks SAF into thinking that network storage (SMB in my case, not sure what other protocols it might work with) is native/device storage and as such, is readily accessible. Very handy app if you use network storage solutions as well at Android devices.

Anyway, thanks again for putting up with my questions. It wasr very helpful, as it kept me thinking about potential solutions until one was arrived at.

And of course, thank you for all you do in developing/maintaining/etc. Cloudstream3! 


PS - Along the same lines, how might one support the project, say, via donation? I thought I found it in the "Give a benene to the devs" "feature" in the app's settings, but unless there's a bug in **4.0.1 • c3296f3**, it would appear that this is more of a "hey thanks" than a financial contribution, as I've given 12 benenes so far and nothing happens other than an additional benene added. I expected it to open a page with donation options but nada.
